### DISCUSSION TITLE
ThinLinc Configuration

### DISCUSSION THREAD

**Discussion by moetiker** [jens-maus/thinRoot#18] [Q&A] [answered]
Hi

Thanks a lot for the cool project. Thinlinc client is running but I can't configure port number and other client settings.

is there a place where I can preset this in a thinlincrc file .... 

found some infos in the script: S42LoadNetFiles . is there any documentation of the concept or some examples files ... In my case tftp server and config web server are not the same host...

**Comment by jens-maus** [comment, marked as answer]
Thanks for asking here on thinRoot specifics. Yes, we are unfortunately lacking quite some documentation and I would be happy to work on a much broader documentation base but I think some help would be required to get this new project going and better distributed. So if you are interested I would be happy to answer all your questions and then you could in fact prepare some more detailed documentation on setup/installation and tuning (e.g. in the wiki section here).

Regarding your question on client-based settings for ThinLinc, etc. All the methods I implemented in thinRoot are largely borrowed from our previous thinstation-based solution (see https://github.com/hzdr/thinstation) which we used as a base for this new project. Thus, all the logic in `/etc/init.d/S42LoadNetFiles` is actually trying to replicate what is usually done in thinstation netfiles on the TFTP server.

So maintaining specific ThinLinc settings per client or client group you can in fact e.g. use the `thinroot.conf.network` file usually located in `/tftpboot/thinroot/conf`(in our case). In our case we are having the following default ThinLinc related settings:

```
SESSION_0_THINLINC_CONFIG_SOUND_ENABLED=1
SESSION_0_THINLINC_CONFIG_REMOVE_CONFIGURATION=0
SESSION_0_THINLINC_CONFIG_SOUND_SYSTEM=AUTO
SESSION_0_THINLINC_CONFIG_NFS_SERVER_ENABLED=1
SESSION_0_THINLINC_CONFIG_NFS_EXPORTS=/run/usbmount,rw
SESSION_0_THINLINC_CONFIG_LOWERCASE_LOGIN_NAME=1
SESSION_0_THINLINC_CONFIG_JPEG_COMPRESSION=0
SESSION_0_THINLINC_CONFIG_HOST_ALIASES=
SESSION_0_THINLINC_CONFIG_OPTIONS_POPUP_KEY=
SESSION_0_THINLINC_CONFIG_ALLOW_HOSTKEY_UPDATE=0
SESSION_0_THINLINC_CONFIG_UPDATE_ENABLED=0
```

All these settings are then upon bootup of thinroot copied into `/home/truser/.thinlinc/tlclient.conf` within the `qutselect-startup.sh` script we are using (see here https://github.com/jens-maus/thinRoot/blob/master/buildroot-external/rootfs-overlay/bin/qutselect-startup.sh#L52-L70). Thus, using this mechanism you should be able to override the default port, etc. in `tlclient.conf` to your likes.

In addition, please see an example  directory tree we are using in our `/tftpboot` directory for thinroot:

```
/tftpboot/thinroot
/tftpboot/thinroot/conf
/tftpboot/thinroot/conf/qutselect.motd
/tftpboot/thinroot/conf/qutselect.slist
/tftpboot/thinroot/conf/thinroot.conf-PC123
/tftpboot/thinroot/conf/thinroot.conf.group-DEPARTMENTA
/tftpboot/thinroot/conf/thinroot.conf.network
/tftpboot/thinroot/conf/thinroot.hosts
/tftpboot/thinroot/overlay
/tftpboot/thinroot/overlay.pkg
/tftpboot/thinroot/overlay_gen.sh
/tftpboot/thinroot/thinroot.img
```

So if you are familiar how tftp-based config files looked in a thinstation setup you should at least get an idea how to set this up with thinRoot as well.

And then in the PXE setup we are using the following settings in a `thinroot.cfg` file under `/tftpboot/pxelinux.cfg` directory:

```
LABEL 	   thinroot
MENU LABEL ^thinRoot (default)
KERNEL     http://X.X.X.X/thinroot/thinroot.img
APPEND     console=tty2 kgdboc=tty2 noswap elevator=deadline consoleblank=120 quiet loglevel=0
TEXT HELP
Our thinRoot Environment
ENDTEXT
```

And as said, don't hesitate to ask questions if you have problems with setting up an example thinRoot installation on your TFTP/PXE server.

**Comment by jens-maus** [comment]
After all these years, please note that I just added a basic demo `/tftpboot` installation directory where you can find all kind of example config/env files you could reuse for your PXE+TFTP/HTTP+DHCP environment to host your thinRoot images. See here:

https://github.com/jens-maus/thinRoot/tree/master/tftpboot

This should hopefully help everyone interested in getting thinRoot usable in their environment to setup their PXE/TFTP/HTTP environment accordingly.

**Comment by swchen1217** [comment]
Hi @jens-maus ,

I'm very glad to have found this project.

I'd like to connect to a PVE VM via SPICE, with the goal of using it as a VDI client.

After booting with generic-x86_64.img, I wasn't able to get into the GUI.

I was wondering if there are any examples of using this as a PVE VDI client?

Thank you very much for your help!

**Reply by jens-maus** [reply]
> I'd like to connect to a PVE VM via SPICE, with the goal of using it as a VDI client.
> 
> After booting with generic-x86_64.img, I wasn't able to get into the GUI.

What exactly is happening or what exactly do you see when booting the generic-x86_64.img via PXE? Are you booting over PXE at all or trying to run the img directly on a harddisk/ssd?
 
> I was wondering if there are any examples of using this as a PVE VDI client?

Unfortunately, this isn't currently so straight forward as it seems. In fact, there should be a `/bin/scripts/qutselect_connect_pve.sh` script that you can use to start/initiate such a SPICE connection to a Proxmox cluster/host. Here, for our VDI client we simply use something like the following environment variable file put on our tftpboot:

```
SESSION_0_XSTARTUP="eval echo PASSWORD | /bin/scripts/qutselect_connect_pve.sh 0 PVE true fullscreen 0 0 0 pvehostname:443 pveuser@pve vdihostname && while pgrep remote-viewer >/dev/null; do sleep 1; done"
```

So have a look for the `/bin/scripts/qutselect_connect_pve.sh` file (cf. https://github.com/hzdr/qutselect/blob/master/scripts/qutselect_connect_pve.sh) and see if this might helps you.

**Reply by swchen1217** [reply]
I used PXE to boot. Here are my screenshots and screen recording.

thinRoot (stable)
<img width="1198" height="664" alt="image" src="https://github.com/user-attachments/assets/41212b35-696b-4651-a8e3-766ae8afab2f" />

thinRoot (stable+debug)
https://github.com/user-attachments/assets/b8f5d8d9-b20d-43bc-a1b1-2f7b4323426d

**Reply by jens-maus** [reply]
It seems that thinroot cannot initialize your local display. What type of hardware is that you are trying to run thinroot on?

**Reply by swchen1217** [reply]
I use PVE VM
<img width="622" height="247" alt="image" src="https://github.com/user-attachments/assets/81817123-6f51-4805-a10b-317ad1dca508" />

**Reply by swchen1217** [reply]
Hi @jens-maus , do you think I should use a physical computer host for testing?

**Reply by jens-maus** [reply]
Yes, I think you should use a physical system (intel NUC or something similarly standard) for testing, because thinroot is/was specifically optimized for physical clients as this is the main purpose and thus is missing virtio drivers and also display drivers as you noticed.

**Reply by swchen1217** [reply]
Hi @jens-maus , I successfully compiled and added the display driver, and can now see the screen. However, after a while, the screen disappears. What should I see normally? Do you have a complete screenshot or video of the operation for me to refer to? Thank you.


https://github.com/user-attachments/assets/a6fcea97-96dc-4496-93aa-7a652651a09b

**Reply by jens-maus** [reply]
Obviously the X11 display driver seems to be still missing because X11 should usually startup. As said, thinRoot's purpose is to be used on physical clients and not on virtual clients as this use-case does not make sense. Or what is your use-case here?

So better grab an intel NUC or other similar intel or amd-graphics chip driven system and try to boot that one instead of trying to get a virtual thinroot running because that simply isn't the target use-case here. But which display driver did you add? Please provide a diff/pull request accordingly.

**Reply by swchen1217** [reply]
These are the results from my test using an ASUS E500G9 PC. After booting in normal mode, the "HZDR" logo appears, and then the progress bar completes, the screen goes completely black with only a flashing "-".

The following are screenshots from my debug mode boot.
<img width="1762" height="991" alt="image" src="https://github.com/user-attachments/assets/cfb63f04-7163-4d0a-a49a-b8bbab68386e" />

Is there any tutorial telling me how to use overlay and conf to set the program to be executed after booting?

**Reply by jens-maus** [reply]
This looks like the same: the graphics card in your E500G9 (Nvidia?) is currently not supported by ThinRoot or the compiled kernel and X11 environment. As said earlier, thinroot is targeted for more **Thin**-Client type of workstations, so something like an intel NUC with integrated graphics cards and not a full-fledged workstations with zillian of different graphics cards possibilities. However, it also should run on a DELL Laptop for example - we have some running thinroot perfectly fine. Please see the list of currently supported hardware on the main page. This even includes different types of RaspberryPis since they are also capable of being used as an appropriate thinclient.

However, I understand that it should at least come up with some more verbose error handling and probably always end up in a login console (Have you tried ALT+F2 to switch to a different one?) so that you can investigate further.

And unfortunately there isn't currently any tutorial yet apart from the documentation on the main page and the `tftpboot` directory in the repository which should serve as an example point

**Reply by swchen1217** [reply]
The E500G9 CPU uses an Intel Core i7-14700, and the graphics card uses its internal graphics card rather than an external graphics card, similar to a typical NUC.

Are the examples in the tftpboot directory specifically for connecting to the PVE VM SPICE? I need some tips on how to use this project.

Thanks so much for your help.

I used Alt+F2 to successfully enter the following screen, but I havn't default username and password.

<img width="501" height="549" alt="image" src="https://github.com/user-attachments/assets/0290c5b9-3e05-45a7-a9be-4786315d0ed7" />

**Reply by jens-maus** [reply]
Sorry for the delay in answering...

> The E500G9 CPU uses an Intel Core i7-14700, and the graphics card uses its internal graphics card rather than an external graphics card, similar to a typical NUC.

Then X11 should usually be able to bring up the user interface. However, in your case it seems something is still missing. Thus, you would need to investigate by logging in and analyzing the logs, etc.

> Are the examples in the tftpboot directory specifically for connecting to the PVE VM SPICE? I need some tips on how to use this project.

Not for a direct connection to a PVE VM via SPICE. But once the connection GUI (qutselect) comes up you should be able to select/enter the name of your SPICE VM and then you should be able to conenct to it via the selection GUI. And on top of that there are also ways to initialize a direct connection to a PVE VM with SPICE once the connection GUI (qutselect) comes up. But right now you seem to be not at that stage yet.

> Thanks so much for your help.

No problem. I also want to get thinRoot into a stage where it would be more easily applicable in a vanilla/fresh environment.

> I used Alt+F2 to successfully enter the following screen, but I havn't default username and password.

See here:
https://github.com/jens-maus/thinRoot/blob/master/tftpboot/thinroot/overlay/etc/shadow

So, you can use for example `mkpasswd` to generate a password hash and put it into `/tftpboot/thinroot/overlay/etc/shadow` and then afterwards re-run the `overlay_gen.sh` script (see https://github.com/jens-maus/thinRoot/blob/master/tftpboot/thinroot/overlay_gen.sh) from within your `/tftpboot/thinroot` directoy on your PXE server and then you should be able to login with `root` + PASSWORD at least once your PXE booted system is rebooted and loads that `overlay.pkg` file upon bootup.

Note, that within `/tftpboot/thinroot/overlay/` there are more files like the dropbear/SSH keys you can put there as well as adding an own `authorized_keys` file with a SSH public key in there so that you can directly SSH into your booted thinroot station, so that you can debug from there.

And once you are in, have  a look for a `/var/log/Xorg.0.log` and `/var/log/boot.log` to analyze what might be wrong with your X11 startup.
